### DISCUSSION TITLE
The operation couldn’t be completed. (HaishinKit.IORecorder.Error error 3.)

### DISCUSSION THREAD

**Discussion by mkrn** [HaishinKit/HaishinKit.swift#1459] [Q&A]
### Describe the bug

Sometimes if the stream getting trouble connecting in the end I get the 
The operation couldn’t be completed. (HaishinKit.IORecorder.Error error 3.)
Is there any reason or a

### To Reproduce

Unfortunately no steps, but this usually happens when there's issue with connecting on 4g / multiple reconnects

### Expected behavior

Recording to always save no matter what

### Version

master

### Smartphone info.

Apple iPhone 12 Pro Max

**Comment by shogo4405** [comment]
I want to know the content of the error message. You can see the content inside Error.Domain with print(error).

1. This commonly occurs when the microphone is detached, and the app is in the background for 30 seconds or more, especially when the Control Center is open for 30 seconds or longer. This can lead to recording failures. 
   > failedToFinishWriting(error: Optional(Error Domain=AVFoundationErrorDomain Code=-11847 "Operation Interrupted" UserInfo={NSLocalizedRecoverySuggestion=Stop other operations and try again., NSLocalizedDescription=Operation Interrupted, NSUnderlyingError=0x2815224c0 {Error Domain=NSOSStatusErrorDomain Code=-16121 "(null)"}}))
1. When you specify recorder.startRecording([.audio: [], .video: []]) and either the video or audio is missing, recording will fail. The error message in this case is as follows:
   >  failedToFinishWriting(error: nil)

**Comment by mkrn** [comment]
Apparently here's what happened:
Swipe to left to other app => app goes to background => sessionWasInterrupted =>  stop recording
While my app was in background, sessionInterruptionEnded called => started streaming & recording
...something empty recorded,
returned to foreground...
stop recording=> error nil

I fixed the logic in my app, but FYI recording error may happen if recording was running in the background.

**Comment by kenle** [comment]
@shogo4405 I started getting this issue as well.  Perhaps every 1 out of 10 file recordings will get failedToFinishWriting(error: nil), it is very random.  I can't figure out what the issue is since error description is nil.

It is a bit concerning when someone is recording something important for 50 minutes and at the end finds out the video did not save :(

I'll try to look into this myself now to see what conditions create this, and if there is a work around.  But just wanted to notify that it is occuring.

**Reply by shogo4405** [reply]
Regarding failedToFinishWriting(error: nil), I am not aware of any solutions other than the method provided in the example. If this method is used, I would like users to handle it on their side. I look forward to receiving a PR with the fix.

**Comment by al3x88** [comment]
Same here... after 2 hours recording I called stopRecording and the file was 37 min only!!

This is the log happening at the end of the 2 hours:

`failedToFinishWriting(error: Optional(Error Domain=AVFoundationErrorDomain Code=-11800 "L’operazione non è stata completata" UserInfo={NSLocalizedFailureReason=Si è verificato un errore sconosciuto (-16364), NSLocalizedDescription=L’operazione non è stata completata, NSUnderlyingError=0x3027d7ba0 {Error Domain=NSOSStatusErrorDomain Code=-16364 "(null)"}}))`

Is there a way to be alerted when the writer somehow stops to record? @shogo4405 

thanks so much!

**Comment by al3x88** [comment]
Hi @shogo4405 I am facing a lot of troubles with recording. What I noticed is that this error

`failedToFinishWriting(error: Optional(Error Domain=AVFoundationErrorDomain Code=-11800 "L’operazione non è stata completata" UserInfo={NSLocalizedFailureReason=Si è verificato un errore sconosciuto (-16364), NSLocalizedDescription=L’operazione non è stata completata, NSUnderlyingError=0x3027d7ba0 {Error Domain=NSOSStatusErrorDomain Code=-16364 "(null)"}}))`

gets triggered only when I call this at the very end of the recording:

```swift
do {
                                    let outputURL = try await self.recorder.stopRecording()
                               
                                } catch {
                                     print(error)

                                }
```
The resulting file is always created but most of time it's ways shorter. 
I believe the issue is coming from the behavour while recording in background.

I performed huge num of tests and I really can't find a pattern that simulates the issue 100% of the times.

The main problem is that **you are not notified about the issue at that moment in time** and you realize this only when you call `stoRecording()`, maybe after 2 hours of recording.

What do you suggest to do? How about call `stopRecording()` when app goes in background and call `startRecording()` again when enters in foreground?

Is there a way to receive a notification when the corruption occurs, so that I can start a new recording and reduce the data loss window?

**Reply by shogo4405** [reply]
This API is a lightweight wrapper around the iOS standard AVAssetWriter. I don't fully understand all the issues that may occur with AVAssetWriter, and the API behavior cannot surpass its limitations. If additional modifications are necessary, I look forward to your PR.

**Reply by al3x88** [reply]
Hi @shogo4405 I finally figured out and the root cause was that

`await recorder.setMovieFragmentInterval(movieFragmentInterval: 10)`

 was issued before

`try await self.recorder.startRecording(myurl, settings: [ ...`

and this leads to unpredictable errors...

Now the recorder works well but If I simulate a crash the resulting video cannot be recovered... seems like setMovieFragmentInterval is ineffective...

**Reply by shogo4405** [reply]
Yes, I don't have all the details, but I've heard both success stories and reports of issues.
At the time of implementation, I tested it locally by killing the task, and I was able to confirm that the recording remained and was playable.
